src/Core/CustomRedirects/CustomRedirectCollection.cs

```diff
@@ -34,45 +70,168 @@ public int IndexOf(CustomRedirect customRedirect)
+                {
+                    if (!redirect.SkipQueryString)
+                    {
+                        var newUrlParts = redirect.NewUrl.Split(new[] { "?" }, StringSplitOptions.RemoveEmptyEntries);
+                        var newUrlQuery = newUrlParts.Length > 1 ? newUrlParts[1] : null;
+
+                        var originalQueryParsed = HttpUtility.ParseQueryString(urlNotFound.Query);
+                        var targetQueryParsed = newUrlQuery != null
+                            ? HttpUtility.ParseQueryString(newUrlQuery)
+                            : new NameValueCollection();
+
+                        var appendQuery = Merge(targetQueryParsed, originalQueryParsed);
+                        var newUrl = redirect.NewUrl;
+
+                        redirect = new CustomRedirect(redirect)
+                        {
+                            NewUrl = newUrl.Contains("?") ? $"{newUrl}&{appendQuery}" : $"{newUrl}?{appendQuery}"
+                        };
+                    }
+                }
             }
+            else
+            {
+                if (!redirect.WildCardSkipAppend)
+                {
+                    //    // We need to append the 404 to the end of the
+                    //    // new one. Make a copy of the redir object as we
+                    //    // are changing it.
+                    var url = urlNotFound.ToString();
+                    CustomRedirect redirCopy = new CustomRedirect(redirect);
+                    var urlFromRule = UrlStandardizer.Standardize(redirect.OldUrl);
 
-            return foundRedirect;
+                    var append = IsAbsoluteUrl(urlFromRule) ? url.Substring(urlFromRule.Length) : urlNotFound.PathAndQuery.Substring(urlFromRule.Length);
+
+                    if (append != string.Empty)
+                    {
+                        redirCopy.NewUrl = UrlStandardizer.Standardize(redirCopy.NewUrl + append);
+                        return redirCopy;
+                    }
+                }
+            }
+
+            return redirect;
+        }
+
+        private bool IsAbsoluteUrl(/*[NotNull]*/ string url)
+        {
+            if (url == null) throw new ArgumentNullException(nameof(url));
+
+            return url.StartsWith("//", StringComparison.InvariantCultureIgnoreCase)
+                   || url.StartsWith("http://", StringComparison.InvariantCultureIgnoreCase)
+                   || url.StartsWith("https://", StringComparison.InvariantCultureIgnoreCase);
+        }
+
+        /*[NotNull]*/
+        private string Merge(/*[NotNull]*/ NameValueCollection targetQuery, /*[NotNull]*/ NameValueCollection originalQuery)
+        {
+            if (targetQuery == null) throw new ArgumentNullException(nameof(targetQuery));
+            if (originalQuery == null) throw new ArgumentNullException(nameof(originalQuery));
+
+            var appendQueryArray = originalQuery.AllKeys.Where(x => targetQuery[x] == null)
```

Comment:
is `Where` predicate correct? You are getting keys out of `originalQuery` collection only when `targetQuery` collection has no item under the same key? If so - does method name describes what it does?

Reply:
In my understanding - it does the merge. Please check if newly added comments shed some light on it. Here is the relevant part of test reports:
  describe CustomRedirectCollection query string merge
      Given the redirect rule is http://mysite/test => http://mysite?a=1&b=2, exactMatch=True, skipQueryString=False
        when url is 'http://mysite/test?a=0&x=0' then it's redirected to 'http://mysite?a=1&b=2&x=0' (1ms)
        when url is 'http://mysite/test/?x=0' then it's redirected to 'http://mysite?a=1&b=2&x=0' (0ms)